rclpy/rclpy/node.py

```diff
@@ -401,42 +412,63 @@ def declare_parameters(
                 )
 
             value = None
-            descriptor = ParameterDescriptor()
+            param_type = None
 
             # Get the values from the tuple, checking its types.
             # Use defaults if the tuple doesn't contain value and / or descriptor.
-            try:
-                name = parameter_tuple[0]
-                assert \
-                    isinstance(name, str), \
-                    (
-                        'First element {name} at index {index} in parameters list '
-                        'is not a str.'.format_map(locals())
-                    )
-
-                # Get value from parameter overrides, of from tuple if it doesn't exist.
-                if not ignore_override and name in self._parameter_overrides:
-                    value = self._parameter_overrides[name].value
-                else:
-                    # This raises a TypeError if it's not possible to get a type from the tuple.
-                    value = parameter_tuple[1]
-
-                # Get descriptor from tuple.
-                descriptor = parameter_tuple[2]
-                assert \
-                    isinstance(descriptor, ParameterDescriptor), \
-                    (
-                        'Third element {descriptor} at index {index} in parameters list '
-                        'is not a ParameterDescriptor.'.format_map(locals())
-                    )
-            except AssertionError as assertion_error:
-                raise TypeError(assertion_error)
-            except IndexError:
-                # This means either value or descriptor were not defined which is fine.
-                pass
+            name = parameter_tuple[0]
+            second_arg = parameter_tuple[1] if 1 < len(parameter_tuple) else None
+            descriptor = parameter_tuple[2] if 2 < len(parameter_tuple) else ParameterDescriptor()
+
+            if not isinstance(name, str):
+                raise TypeError(
+                        f'First element {name} at index {index} in parameters list '
+                        'is not a str.')
+            if not isinstance(descriptor, ParameterDescriptor):
+                raise TypeError(
+                    f'Third element {descriptor} at index {index} in parameters list '
+                    'is not a ParameterDescriptor.'
+                )
+
+            if len(parameter_tuple) == 1:
+                warnings.warn(
+                    f"when declaring parmater named '{name}', "
+                    'declaring a parameter only providing its name is deprecated. '
+                    'You have to either:\n'
+                    '\t- Pass a name and a default value different to "PARAMETER NOT SET"'
+                    ' (and optionally a descriptor).\n'
+                    '\t- Pass a name and a parameter type.\n'
+                    '\t- Pass a name and a descriptor with `dynamic_typing=True',
+                    DeprecationWarning)
```

Comment:
Sounds good, but I think we should start following python usual practices instead of having our own unusual convention.

Here's a pep that have some of the history of why `DeprecationWarning` is only enable by default for `__main__`: https://www.python.org/dev/peps/pep-0565/.

If we want to catch deprecation warnings in our codebase, we can use `python -X dev` (developer mode), when running tests.

Reply:
I'm less concerned about us catching deprecation warnings and more-so about users missing them.

I just tried adding the following line to [this example](https://github.com/ros2/examples/blob/54b437c551eb15a5579d7e58d01f8446b8de6540/rclpy/topics/minimal_subscriber/examples_rclpy_minimal_subscriber/subscriber_lambda.py):

```python
node.declare_parameter('foo')
```

According to PEP 565, I expected to see a deprecation warning when running like this:

```
python3 subscriber_lambda.py
```

But I don't see any warning... I'm not sure what the expected behaviour is in this case; it seems like we're executing via `__main__`.

Reply:
> I just tried adding the following line to this example:

did you add that line to the main() function?

iiuc the idea of pep565 is that the warning is only shown if you call the function directly in `__main__`, and not through another function (in this case the `main()` function).

My understanding is that an end user cannot do much about a deprecation warning, and the idea is to only show it "for developers".

Reply:
Done in https://github.com/ros2/rclpy/pull/683/commits/9db99802b70d1cb03ea32ed1401bda95daf6c3ac.

Reply:
I just added the line to `main()`, I guess that's why I didn't see anything then.